### PR TITLE
Improve testbed template construction

### PR DESCRIPTION
**Pull request by touilleMan**

### CODE REVIEW

File: .cspell/custom-words.txt [hunk 1]
```diff
@@ -214,6 +214,7 @@ lproj
 lsregister
 lstfiracode
 lualatex
+Mcalice
```
[the remark this answers is not in the document]
Comment (touilleMan) [reply .cspell/custom-words.txt:217]: done.

File: libparsec/crates/testbed/src/templates/mod.rs [hunk 2]
```diff
@@ -1,6 +1,7 @@
 // Parsec Cloud (https://parsec.cloud) Copyright (c) BUSL-1.1 (eventually AGPL-3.0) 2016-present Scille SAS
 
-pub(crate) mod alice_cruising;
+// pub(crate) mod alice_cruising;
```
Comment (FirelightFlagboy) [inline libparsec/crates/testbed/src/templates/mod.rs:3]: To remove

File: server/packaging/server/in-docker-build.sh [hunk 3]
```diff
@@ -7,18 +7,33 @@ sed -i -e '\;bindings;d' Cargo.toml
 
 # Install Rust (actual toolchain is going to be installed by maturin according to `rust-toolchain.toml`)
 curl https://sh.rustup.rs -sSf | sh -s -- -y --default-toolchain none
-
 export PATH="/root/.cargo/bin:$PATH"
+cargo --version
+
+# Install Poetry
+curl -sSL https://install.python-poetry.org | python
```
Comment (FirelightFlagboy) [inline server/packaging/server/in-docker-build.sh:14]: Could you consider fixing the installed poetry version ?

File: server/packaging/testbed-server/build.sh [hunk 4]
```diff
@@ -13,14 +13,14 @@ CURRENT_VERSION=$(grep -o '^version = .*$' $ROOTDIR/server/pyproject.toml | sed
...
+PREFIX=ghcr.io/scille/parsec-cloud
 
 echo "Will create an image \`parsec-testbed-server\` with the following tags:"
 echo "- \`latest\`"
 echo "- \`$UNIQ_TAG\`"
 
-PREFIX=ghcr.io/scille/parsec-cloud
-
-if [ "$WRITE_ENV" -eq 1 ]; then
+if [ "$WRITE_ENV" = "1" ]; then
```
Comment (FirelightFlagboy) [inline server/packaging/testbed-server/build.sh:23]: Why the change ?

File: server/packaging/testbed-server/in-docker-build.sh [hunk 5]
```diff
@@ -7,18 +7,34 @@ sed -i -e '\;bindings;d' Cargo.toml
 
 # Install Rust (actual toolchain is going to be installed by maturin according to `rust-toolchain.toml`)
 curl https://sh.rustup.rs -sSf | sh -s -- -y --default-toolchain none
-
 export PATH="/root/.cargo/bin:$PATH"
+cargo --version
+
+# Install Poetry
+curl -sSL https://install.python-poetry.org | python
```
Comment (FirelightFlagboy) [inline server/packaging/testbed-server/in-docker-build.sh:14]: Could you consider fixing poetry version ?

File: server/packaging/testbed-server/build.sh (hunk 4 above)
Comment (touilleMan) [reply server/packaging/testbed-server/build.sh:23]: if WRITE_ENV is empty, `if [ "" -eq 1 ]` is not valid

File: libparsec/crates/client/src/user_ops/mod.rs [hunk 6]
```diff
@@ -73,6 +73,11 @@ impl UserOps {
         self.storage.stop().await;
     }
 
+    /// Low-level access, should be only needed for tests
+    pub fn test_get_user_manifest(&self) -> Arc<LocalUserManifest> {
+        self.storage.get_user_manifest()
+    }
```
[the remark this answers is not in the document]
Comment (touilleMan) [reply libparsec/crates/client/src/user_ops/mod.rs:79]: That was my idea at first, but it was not possible to do given it is `libparsec_client`'s test that require to access them (and you cannot specify feature for the crate currently tested)

The alternative is to move all test needing this into the unit tests, but it's cumbersome (this integration vs unit test in Rust is a real pain to deal with :disappointed: )

So in the end I though this is the least worst approach...

File: libparsec/crates/platform_storage/src/testbed.rs [hunk 7]
```diff
@@ -45,14 +45,15 @@ pub(crate) async fn maybe_populate_certificate_storage(data_base_dir: &Path, dev
...
-            let up_to_index = env.template.events().rev().find_map(|e| match e {
+            let up_to_index = env.template.events.iter().rev().find_map(|e| match e {
                 TestbedEvent::CertificatesStorageFetchCertificates(x)
                     if x.device == device.device_id =>
                 {
                     Some(x.up_to_index)
                 }
                 _ => None,
             });
+            println!("============> UP TO INDEX {:?}", &up_to_index);
```
[the remark this answers is not in the document]
Comment (touilleMan) [reply libparsec/crates/platform_storage/src/testbed.rs:56]: :+1: done.

File: libparsec/crates/testbed/src/templates/mod.rs (hunk 2 above)
Comment (touilleMan) [reply libparsec/crates/testbed/src/templates/mod.rs:3]: :+1: done

File: server/packaging/server/in-docker-build.sh (hunk 3 above)
Comment (touilleMan) [reply server/packaging/server/in-docker-build.sh:14]: good idea ! done.

File: server/packaging/testbed-server/in-docker-build.sh (hunk 5 above)
Comment (touilleMan) [reply server/packaging/testbed-server/in-docker-build.sh:14]: done.

File: libparsec/crates/platform_storage/tests/certificates.rs [hunk 8]
```diff
@@ -4,9 +4,60 @@ use libparsec_platform_storage::certificates::CertificatesStorage;
...
+        builder.new_user("bob");
+        builder.new_realm("bob");
+
+        expected_last_index = Some((
+            builder.current_certificate_index(),
+            builder.current_timestamp(),
+        ));
+        builder.certificates_storage_fetch_certificates("alice@dev1");
+
+        if kind == "multiple_fetch" {
```
[the remark this answers is not in the document]
Comment (touilleMan) [reply libparsec/crates/platform_storage/tests/certificates.rs:23]: I find it more readable, especially given the fact that there is exactly 2 kinds is purely coincidental (cf. the tests in `user.rs` for instance)

File: server/packaging/testbed-server/build.sh (hunk 4 above)
Comment (FirelightFlagboy) [reply server/packaging/testbed-server/build.sh:23]: :+1:

File: libparsec/crates/platform_storage/tests/certificates.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (touilleMan) [reply libparsec/crates/platform_storage/tests/certificates.rs:23]: done. (I've also switch to enum in the `user.rs` corresponding test

[the remark this answers is not in the document]
Comment (FirelightFlagboy) [reply libparsec/crates/platform_storage/tests/certificates.rs]: :+1:
